File: androidApp/src/main/java/com/andremw96/notesnotes_kmm/android/ui/login/LoginViewModel.kt
package com.andremw96.notesnotes_kmm.android.ui.login

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.andremw96.notesnotes_kmm.domain.GetCredential
import com.andremw96.notesnotes_kmm.domain.Login
import com.andremw96.notesnotes_kmm.domain.SaveCredential
import com.andremw96.notesnotes_kmm.model.repository.LoginDataValidator
import com.andremw96.notesnotes_kmm.network.utils.Resource
import kotlinx.coroutines.launch

class LoginViewModel(
    private val login: Login,
    private val loginDataValidator: LoginDataValidator,
    private val saveCredential: SaveCredential,
    private val getCredential: GetCredential,
) : ViewModel() {

    private var _loginForm: MutableLiveData<LoginFormState> = MutableLiveData()
    val loginFormState: LiveData<LoginFormState> = _loginForm

    fun checkLogin() {
        viewModelScope.launch {
            val credential = getCredential.invoke()
            if (credential.first != "" && credential.second != "") {
                _loginForm.postValue(
                    LoginFormState(
                        username = credential.second,
                        password = "",
                        email = null,
                        accessToken = credential.first,
                        usernameError = null,
                        passwordError = null,
                        loginError = null,
                        isLoginSuccess = true
                    )
                )
            }
        }
    }

    fun loginDataChanged(username: String, password: String) {
        var usernameError: String? = null
        val checkUsername = loginDataValidator.checkUsername(username)
        if (checkUsername is LoginDataValidator.LoginValidatorResult.Error) {
            usernameError = checkUsername.message
        }

        var pwdError: String? = null
        val checkPwd = loginDataValidator.checkPassword(password)
        if (checkPwd is LoginDataValidator.LoginValidatorResult.Error) {
            pwdError = checkPwd.message
        }

        _loginForm.postValue(
            LoginFormState(
                username = username,
                password = password,
                usernameError = usernameError,
                passwordError = pwdError,
            )
        )
    }

    fun login(username: String, password: String) {
        viewModelScope.launch {
            _loginForm.postValue(
                LoginFormState(
                    username = username,
                    password = password,
                    isLoading = true,
                    isLoginSuccess = false
                )
            )

            if (_loginForm.value?.isDataValid == true) {
                val login = login.invoke(username, password)
                if (login is Resource.Success) {
                    _loginForm.postValue(
                        LoginFormState(
                            username = username,
                            password = password,
                            email = login.data?.user?.email,
                            accessToken = login.data?.accessToken,
                            usernameError = null,
                            passwordError = null,
                            loginError = null,
                            isLoginSuccess = true
                        )
                    )
                    login.data?.accessToken?.let {
                        saveCredential.invoke(username, it)
                    }
                } else {
                    _loginForm.postValue(
                        LoginFormState(
                            username = username,
                            password = password,
                            email = null,
                            accessToken = null,
                            usernameError = null,
                            passwordError = null,
                            loginError = login.message,
                            isLoginSuccess = false
                        )
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/Screens/Login/LoginViewModel.swift
//
//  LoginViewModel.swift
//  iosApp
//
//  Created by Quipper Indonesia on 16/12/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared


@MainActor class LoginViewModel: ObservableObject {
    @Published var viewState: LoginViewState = LoginViewState()
    
    private let loginDataValidator: LoginDataValidator
    private let loginUseCase: Login
    private let saveCredential: SaveCredential
    private let getCredential: GetCredential
    
    init(loginDataValidator: LoginDataValidator, loginUseCase: Login, saveCredential: SaveCredential, getCredential: GetCredential) {
        self.loginDataValidator = loginDataValidator
        self.loginUseCase = loginUseCase
        self.saveCredential = saveCredential
        self.getCredential = getCredential
    }
    
    func validateUsernamePassword(username: String, password: String) {
        viewState = LoginViewState(
            username: username,
            password: password,
            usernameError: (loginDataValidator.checkUsername(username: username) as? LoginDataValidator.LoginValidatorResultError)?.message ?? "",
            passwordError: (loginDataValidator.checkPassword(password: password) as? LoginDataValidator.LoginValidatorResultError)?.message ?? ""
        )
    }
    
    func checkLogin(onAlreadyLogin: () -> Void) async {
        do {
            let credential = try await getCredential.invoke()
            if (credential.accessToken != "" && credential.username != "" && credential.userid != -1) {
                onAlreadyLogin()
            }
        } catch {
            print(error)
        }
    }
    
    func login(username: String, password: String, onLoginSuccess: () -> Void) async -> Void {
        viewState = LoginViewState(
            username: username,
            password: password,
            usernameError: "",
            passwordError: "",
            isLoading: true
        )
        
        do {
            let response = try await loginUseCase.invoke(username: username, password: password)
            
            if (response is ResourceSuccess) {
                viewState = LoginViewState(
                    username: response.data?.user.username ?? username,
                    password: password,
                    usernameError: "",
                    passwordError: "",
                    isLoading: false,
                    accessToken: response.data?.accessToken ?? "",
                    loginError: "",
                    isLoginSuccess: true
                )
                
                if response.data != nil {
                    try await saveCredential.invoke(username: username, token: response.data!.accessToken, userId: (response.data?.user.userId)!)
                    onLoginSuccess()
                }
            } else {
                viewState = LoginViewState(
                    username: username,
                    password: password,
                    usernameError: "",
                    passwordError: "",
                    isLoading: false,
                    accessToken: "",
                    loginError: response.message ?? "something went wrong, please try again",
                    isLoginSuccess: false
                )
            }
        } catch {
            print(error)
            
            viewState = LoginViewState(
                username: username,
                password: password,
                usernameError: "",
                passwordError: "",
                isLoading: false,
                accessToken: "",
                loginError: error.localizedDescription,
                isLoginSuccess: false
            )
        }
    }
}
